lisa/schema.py

```diff
@@ -353,88 +355,355 @@ def __post_init__(self, *args: Any, **kwargs: Any) -> None:
+        default=constants.ENVIRONMENTS_NODES_REQUIREMENT,
         metadata=metadata(
-            required=True, validate=validate.OneOf([constants.ENVIRONMENTS_NODES_SPEC]),
+            required=True,
+            validate=validate.OneOf([constants.ENVIRONMENTS_NODES_REQUIREMENT]),
         ),
     )
     name: str = ""
     is_default: bool = field(default=False)
     # optional, if there is only one artifact.
     artifact: str = field(default="")
-    core_count: int = field(
-        default=1,
+    node_count: search_space.CountSpace = field(
+        default=search_space.IntRange(min=1), metadata=metadata(data_key="nodeCount"),
+    )
+    core_count: search_space.CountSpace = field(
+        default=search_space.IntRange(min=1),
         metadata=metadata(data_key="coreCount", validate=validate.Range(min=1)),
     )
-    memory_gb: int = field(
-        default=1,
-        metadata=metadata(data_key="memoryGb", validate=validate.Range(min=1)),
+    memory_mb: search_space.CountSpace = field(
+        default=search_space.IntRange(min=512),
+        metadata=metadata(data_key="memoryMb", validate=validate.Range(min=512)),
+    )
+    nic_count: search_space.CountSpace = field(
+        default=search_space.IntRange(min=1),
+        metadata=metadata(data_key="nicCount", validate=validate.Range(min=1)),
     )
-    gpu_count: int = field(
-        default=0,
+    gpu_count: search_space.CountSpace = field(
+        default=None,
         metadata=metadata(data_key="gpuCount", validate=validate.Range(min=0)),
     )
+    # all features on requirement should be included.
+    # all features on capability can be included.
+    features: Optional[search_space.SetSpace[Feature]] = field(
+        default=None, metadata=metadata()
+    )
+    # set by requirements
+    # capability's is ignored
+    excluded_features: Optional[search_space.SetSpace[Feature]] = field(
+        default=None, metadata=metadata(data_key="excludedFeatures")
+    )
 
+    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
+        if self.features:
+            self.features.is_allow_set = True
+        if self.excluded_features:
+            self.excluded_features.is_allow_set = False
+
+    def __eq__(self, other: object) -> bool:
+        assert isinstance(other, NodeSpace), f"actual: {type(other)}"
+
+        result = (
+            self.node_count == other.node_count
+            and self.core_count == other.core_count
+            and self.memory_mb == other.memory_mb
+            and self.nic_count == other.nic_count
+            and self.gpu_count == other.gpu_count
+        )
+        return result
+
+    def __repr__(self) -> str:
+        return (
+            f"count:{self.node_count}, core:{self.core_count}, "
+            f"mem:{self.memory_mb}, nic:{self.nic_count}, gpu:{self.gpu_count}, "
+            f"f:{self.features}, ef:{self.excluded_features}"
+        )
+
+    def meet_result(self, capability: Any) -> search_space.RequirementResult:
```

Comment:
What's this mean, "meet_result"?

Reply:
Check if capability meet the requirement. The result includes explainable reasons, it can help understand why capability doesn't meet.

Reply:
changed method name to `check`